drivers: lora: Add support for setting sync-word (public/private network) and iq-inverted in lora_modem_config

timcooijmans opened the pull request:
Some use-cases of sending LoRa packets require sending with the public-network sync-word and/or with iq-inverted.
Additionally document the lora_modem_config struct.

Signed-off-by: Tim Cooijmans <timcooijmans@gmail.com>

JordanYates commented on include/zephyr/drivers/lora.h:
```diff
@@ -46,13 +46,36 @@ enum lora_coding_rate {
 };
 
 struct lora_modem_config {
+	/* Frequency in Hz to use for transceiving */
```
Nitpick: Zephyr prefers each commit to do a single thing.
Adding documentation for the existing parameters and adding the new ones should be 2 seperate commits.

JordanYates commented on include/zephyr/drivers/lora.h:
```diff
@@ -46,13 +46,36 @@ enum lora_coding_rate {
+	/* Set to true for transmission, false for receiving */
 	bool tx;
+
+	/* Invert the IQ. Usually can be set to false, but should be set to true for sending a downlink. */
```
What situation is this comment referring to? LoRa doesn't have a "downlink" concept.
It's fine to mention that this is true for LoRaWAN's PHY layer on downlinks only, but it shouldn't be implied that this is somehow a convention for LoRa itself.

JordanYates commented on include/zephyr/drivers/lora.h:
```diff
@@ -46,13 +46,36 @@ enum lora_coding_rate {
+	/* Sets the sync-byte to use: 
+	 * true for using the public network sync byte
+	 * false for using the private network sync byte
+	 * WARNING: Only set to true when sending LoRaWAN packets
```
I am wary of this line, as it implies that the LoRa API should be used for LoRaWAN packets.
Normally users should be directed to the LoRaWAN API if they wish to send LoRaWAN packets.
Mentioning that LoRaWAN uses the public network sync byte is fine.

JordanYates commented on include/zephyr/drivers/lora.h:
```diff
@@ -46,13 +46,36 @@ enum lora_coding_rate {
+	 * false for using the private network sync byte
+	 * WARNING: Only set to true when sending LoRaWAN packets
+	 */
+	bool public_network;
```
These two new parameters should also be explicitly set in the 2 LoRa samples (`samples/drivers/lora`) in a third commit.

JordanYates reviewed (commented):
Thanks for the contributions, especially the documentation on the existing code.
There are a few implications in the comments that I think need to be addressed, but I don't see any problems with the functionality.

timcooijmans committed:
drivers: lora: Add documentation for the lora_modem_config struct.

Adds documentation for the fields in lora_modem_config.

Signed-off-by: Tim Cooijmans <timcooijmans@gmail.com>

timcooijmans committed:
drivers: lora: Add support for setting sync-word and iq-inverted

Add support for setting sync-word and iq-inverted in lora_modem_config. By
being able to set the sync-word it's possible to send "raw" public-network
packets using lora_send and using iq-inverted both uplink and downlink
messages can be sent.

Signed-off-by: Tim Cooijmans <timcooijmans@gmail.com>

timcooijmans commented:
@JordanYates tried to resolve most of the comments, will add the additions to the samples once we agree on this as it will be another commit.

timcooijmans commented:
@JordanYates how can we get this merged?

JordanYates commented on include/zephyr/drivers/lora.h:
```diff
@@ -66,6 +66,21 @@ struct lora_modem_config {
+	/* Invert the IQ. Normally this should be set to false. In advanced
+	 * use-cases where a differentation is needed between "uplink" and
+	 * "downlink" traffic, the IQ can be inverted to create two different
+	 * channels on the same frequency */
```
terminating `*/` on new line

JordanYates commented on include/zephyr/drivers/lora.h:
```diff
@@ -66,6 +66,21 @@ struct lora_modem_config {
 	/* Set to true for transmission, false for receiving */
 	bool tx;
+
+	/* Invert the IQ. Normally this should be set to false. In advanced
```
It might be worthwhile being explicit with In-Phase and Quadrature naming.
`Invert the In-Phase and Quadrature (IQ) signals.` That would give people a better starting point to googling additional info

JordanYates commented on include/zephyr/drivers/lora.h:
```diff
@@ -66,6 +66,21 @@ struct lora_modem_config {
+	bool iq_inverted;
+
+	/* Sets the sync-byte to use:
+	 * false for using the private network sync byte
```
Maybe indent this line and the next as you appear to be creating a list

JordanYates commented on include/zephyr/drivers/lora.h:
```diff
@@ -66,6 +66,21 @@ struct lora_modem_config {
+	/* Sets the sync-byte to use:
+	 * false for using the private network sync byte
+	 * true for using the public network sync byte
+	 * true is only intended for advanced usage. Normally false should be used
```
I would document this in terms of public/private not true/false. For example:
`The public network sync-byte is only intended for advanced usage.`

JordanYates reviewed (changes requested):
Sorry for the delay, much cleaner overall, a few remaining comments.

timcooijmans committed:
samples: drivers: lora: Add public_network and iq_inverted config

Explicitly set the newly add public_network and iq_inverted settings
in lora_modem_config in the lora send and receive samples.

Signed-off-by: Tim Cooijmans <timcooijmans@gmail.com>

timcooijmans commented:
Thank you @JordanYates I also modified the samples now.

JordanYates commented on include/zephyr/drivers/lora.h:
```diff
@@ -61,11 +61,31 @@ struct lora_modem_config {
 	/* Length of the preamble */
 	uint16_t preamble_len;
 
-	/* TX-power in dBm  to use for transmission */
```
This change should be in the first commit of this PR (that added the comment)

JordanYates reviewed (changes requested):
One last minor nit. I think the first commit will also need some body text to pass CI.
